pkg/querier/batch/stream.go

```diff
@@ -174,7 +212,22 @@ func (bs *batchStream) merge(batch *chunk.Batch, size int) {
 					_, fh := batch.AtFloatHistogram()
 					bs.fhPool.Put((*histogram.FloatHistogram)(fh))
 				}
+			} else {
+				populate(batch, rt)
+
+				// if bs.hPool is not nil, we put there the discarded histogram.Histogram object from batch, so it can be reused.
+				if lt == chunkenc.ValHistogram && bs.hPool != nil {
+					_, h := bs.curr().AtHistogram()
+					bs.hPool.Put((*histogram.Histogram)(h))
+				}
+				// if bs.fhPool is not nil, we put there the discarded histogram.FloatHistogram object from batch, so it can be reused.
+				if lt == chunkenc.ValFloatHistogram && bs.fhPool != nil {
+					_, fh := bs.curr().AtFloatHistogram()
+					bs.fhPool.Put((*histogram.FloatHistogram)(fh))
+				}
```

Comment:
@krajorama Do you know why this wasn't done before? Is it unsafe doing it?

Reply:
Not sure what you mean, we never had a situation before where we picked the right side and the other side wasn't a float which is _not_ pooled.